src/ssd/ssd_shadow.c

```diff
@@ -0,0 +1,279 @@
+		break;
+	case LAB_SSD_PART_CORNER_BOTTOM_LEFT:
+		x = -visible_shadow_width;
+		y = -titlebar_height + height - inset + overlap_px;
+		wlr_scene_node_set_position(part->node, x, y);
+		corner_scale_crop(scene_buf, overlap_px, corner_size, true);
+		break;
+	case LAB_SSD_PART_CORNER_TOP_LEFT:
+		x = -visible_shadow_width;
+		y = -titlebar_height - visible_shadow_width;
+		wlr_scene_node_set_position(part->node, x, y);
+		corner_scale_crop(scene_buf, overlap_px, corner_size, false);
+		break;
+	case LAB_SSD_PART_CORNER_TOP_RIGHT:
+		x = width - inset;
+		y = -titlebar_height - visible_shadow_width;
+		wlr_scene_node_set_position(part->node, x, y);
+		corner_scale_crop(scene_buf, overlap_px, corner_size, true);
+		break;
+	case LAB_SSD_PART_RIGHT:
+		x = width;
+		y = -titlebar_height + inset - overlap_px;
+		wlr_scene_node_set_position(part->node, x, y);
+		wlr_scene_buffer_set_dest_size(
+			scene_buf, visible_shadow_width, height - 2 * inset);
+		wlr_scene_node_set_enabled(part->node, show_sides);
+		break;
+	case LAB_SSD_PART_BOTTOM:
+		x = inset;
+		y = -titlebar_height + height;
+		wlr_scene_node_set_position(part->node, x, y);
+		wlr_scene_buffer_set_dest_size(
+			scene_buf, width - 2 * inset, visible_shadow_width);
+		break;
+	case LAB_SSD_PART_LEFT:
+		x = -visible_shadow_width;
+		y = -titlebar_height + inset - overlap_px;
+		wlr_scene_node_set_position(part->node, x, y);
+		wlr_scene_buffer_set_dest_size(
+			scene_buf, visible_shadow_width, height - 2 * inset);
+		wlr_scene_node_set_enabled(part->node, show_sides);
+		break;
+	case LAB_SSD_PART_TOP:
+		x = inset;
+		y = -titlebar_height - visible_shadow_width;
+		wlr_scene_node_set_position(part->node, x, y);
+		wlr_scene_buffer_set_dest_size(
+			scene_buf, width - 2 * inset, visible_shadow_width);
+		break;
+	default:
+		break;
+	}
+}
+
+static void
+set_shadow_geometry(struct ssd *ssd)
+{
+	struct view *view = ssd->view;
+	struct theme *theme = view->server->theme;
+	int titlebar_height = ssd->titlebar.height;
+	int width = view->current.width;
+	int height = view_effective_height(view, false) + titlebar_height;
+
+	struct ssd_part *part;
+	struct ssd_sub_tree *subtree;
+
+	FOR_EACH_STATE(ssd, subtree) {
+		bool active = subtree == &ssd->shadow.active;
+		int visible_shadow_width = active ? theme->window_active_shadow_radius
+			: theme->window_inactive_shadow_radius;
+		/* inset as a proportion of shadow width */
+		double inset_proportion = active ? theme->window_active_shadow_inset
+			: theme->window_inactive_shadow_inset;
+		/* inset in actual pixels */
+		int inset = inset_proportion * (double)visible_shadow_width;
+
+		/*
+		 * Total size of corner buffers including inset and visible
+		 * portion.  Top and bottom are the same size (only the cutout
+		 * is different).  The buffers are square so width == height.
+		 */
+		int corner_size = active
+			? theme->shadow_corner_top_active->unscaled_height
+			: theme->shadow_corner_top_inactive->unscaled_height;
```

Comment:
This crashes labwc if buffer allocation for any shadow part fails. For example,
```
window.active.shadow.inset: 0
window.active.shadow.radius: 0
```
will cause labwc to segfault the first time a window is drawn. The shadow logic needs to handle these cases more gracefully, see comments in `create_shadows` below.

Reply:
Fixed